src/extension/declarative.rs

```diff
@@ -0,0 +1,293 @@
+        let mut ext = Extension::new_with_reqs(self.name.clone(), imports.clone());
+
+        for t in &self.types {
+            t.register(&mut ext, scope, registry)?;
+        }
+
+        for o in &self.operations {
+            o.register(&mut ext, scope, registry)?;
+        }
+
+        Ok(ext)
+    }
+}
+
+/// Errors that can occur while loading an extension set.
+#[derive(Debug, thiserror::Error)]
+pub enum ExtensionDeclarationError {
+    /// An error occurred while deserializing the extension set.
+    #[error("Error while parsing the extension set yaml: {0}")]
+    Deserialize(#[from] serde_yaml::Error),
+    /// An error in registering the loaded extensions.
+    #[error("Error registering the extensions.")]
+    ExtensionRegistryError(#[from] ExtensionRegistryError),
+    /// An error occurred while adding operations or types to the extension.
+    #[error("Error while adding operations or types to the extension: {0}")]
+    ExtensionBuildError(#[from] ExtensionBuildError),
+    /// Invalid yaml file.
+    #[error("Invalid yaml declaration file {0}")]
+    InvalidFile(#[from] std::io::Error),
+    /// A required extension is missing.
+    #[error("Missing required extension {ext}")]
+    MissingExtension {
+        /// The missing imported extension.
+        ext: ExtensionId,
+    },
+    /// Referenced an unknown type.
+    #[error("Extension {ext} referenced an unknown type {ty}.")]
+    MissingType {
+        /// The extension that referenced the unknown type.
+        ext: ExtensionId,
+        /// The unknown type.
+        ty: TypeName,
+    },
+    /// Parametric types are not currently supported as type parameters.
+    ///
+    /// TODO: Support this.
+    #[error("Found a currently unsupported higher-order type parameter {ty} in extension {ext}")]
+    ParametricTypeParameter {
+        /// The extension that referenced the unsupported type parameter.
+        ext: ExtensionId,
+        /// The unsupported type parameter.
+        ty: TypeName,
+    },
+    /// Parametric types are not currently supported as type parameters.
+    ///
+    /// TODO: Support this.
+    #[error("Found a currently unsupported parametric operation {op} in extension {ext}")]
+    ParametricOperation {
+        /// The extension that referenced the unsupported op parameter.
+        ext: ExtensionId,
+        /// The operation.
+        op: SmolStr,
+    },
+    /// Operation definitions with no signature are not currently supported.
+    ///
+    /// TODO: Support this.
+    #[error(
+        "Operation {op} in extension {ext} has no signature. This is not currently supported."
+    )]
+    MissingSignature {
+        /// The extension containing the operation.
+        ext: ExtensionId,
+        /// The operation with no signature.
+        op: SmolStr,
+    },
+    /// Since currently we hard-code some basic types, we require the prelude to be in scope when using them.
+    ///
+    /// TODO: Don't require this.
```

Comment:
? the prelude is expected to be in scope right?

Reply:
Right. I auto-included it.